core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -417,7 +417,7 @@ private[log] class Cleaner(val id: Int,
     txnIndexFile.delete()
 
     val startOffset = segments.head.baseOffset
-    val records = FileRecords.open(logFile, false, log.initFileSize(), log.config.preallocate)
+    val records = FileRecords.open(logFile, false, log.initFileSize(), false)
```

Comment:
Won't this lead to a fragmented file still?  I'd think we'd want to preallocate this file (if enabled), and then remove all of the invalid bytes to shrink it down as is done in segment recovery.